### DISCUSSION TITLE
v0.6.8 Experimental Nerf, Deep Edge Detection, community demos, Quaterion slerp, augmentations dispatcher, image matching and other bug fixes

### DISCUSSION THREAD

**Discussion by edgarriba** [kornia/kornia#1945] [Releases]
# Highlights

## NeRF API

In this release in we include an experimental `kornia.nerf` submodule with a high level API that implements a vanilla Neural Radiance Field (NeRF). Read more about the roadmap of this project: https://github.com/kornia/kornia/issues/1936 // contribution done by @YanivHollander 

```python
from kornia.nerf import NerfSolver
from kornia.geomtry.camera import PinholeCamera

 camera: PinholeCamera = create_one_camera(5, 9, device, dtype)
 img = create_red_images_for_cameras(camera, device)

 nerf_obj = NerfSolver(device=device, dtype=dtype)
 num_img_rays = 15
 nerf_obj.init_training(camera, 1.0, 3.0, False, img, num_img_rays, batch_size=5, num_ray_points=10, lr=1e-2)
 nerf_obj.run(num_epochs=10)

 img_rendered = nerf_obj.render_views(camera)[0].permute(2, 0, 1)
```
![ezgif com-gif-maker](https://user-images.githubusercontent.com/5157099/195349208-72351132-8e73-4110-adb3-466530529734.gif)

Improvements, docs and tutorials soon!

##  Edge Detection

Added `kornia.contrib.EdgeDetection` API that implements `dexined`:  https://github.com/xavysp/DexiNed

```python
import kornia as K
from kornia.contrib import EdgeDetection

edge_detection = EdgeDetector().to(device)

# preprocess
img = K.image_to_tensor(frame, keepdim=False).to(device)
img = K.color.bgr_to_rgb(img.float())

# detect !
with torch.no_grad():
    edges = edge_detection(img)

img_vis = K.tensor_to_image(edges.byte())
```
![amiga_edge](https://user-images.githubusercontent.com/5157099/195353178-3615507e-6b95-4efe-9b26-9ff7a078204d.png)

## Image matching bugfixes:

After testing kornia LoFTR and AdaLAM under big load, our users and we have experiences some bugs in corners cases, such as big images or no input correspondences, which caused pipeline to crash. Not anymore!

* Fixes typo bug that influences LoFTR training by @georg-bn in https://github.com/kornia/kornia/pull/1854
* Enlargen LoFTR positional encoding map if large images are input by @georg-bn in https://github.com/kornia/kornia/pull/1853
* Make AdaLAM output match confidence by @ducha-aiki in https://github.com/kornia/kornia/pull/1862
* fix AdaLAM crash by @ducha-aiki in https://github.com/kornia/kornia/pull/1881
* Adalam fix2 by @ducha-aiki in https://github.com/kornia/kornia/pull/1888
* No crash in local feature matching if empty tensor output by @ducha-aiki in https://github.com/kornia/kornia/pull/1890
* Fix warning in AdaLAM by @Skydes in https://github.com/kornia/kornia/pull/1925

## Various kornia demos in gradio by community:

See demos in our HuggingFace space:  https://huggingface.co/kornia
<img width="887" alt="image" src="https://user-images.githubusercontent.com/4803565/195351434-85a52d81-da28-47db-9915-c9e621316506.png">

* Added gradio Image Stitching demo link by @kadirnar in https://github.com/kornia/kornia/pull/1871
* edge detection demo by @p-mishra1 in https://github.com/kornia/kornia/pull/1876
* Added Hugging Face edge detection demo link by @ramon-rd in https://github.com/kornia/kornia/pull/1874
* [docs] add gradio app html and embeddings in filters by @lappemic in https://github.com/kornia/kornia/pull/1883
* Geometry image transform demo by @dvando in https://github.com/kornia/kornia/pull/1922
* add spaces demo by @johko in https://github.com/kornia/kornia/pull/1905
* add space demo for homography warping by @johko in https://github.com/kornia/kornia/pull/1924
* created resize_antialias.html file by @gauthamk28 in https://github.com/kornia/kornia/pull/1877
* I added html file for module Line Fitting by @kadirnar in https://github.com/kornia/kornia/pull/1886
* Add edge detector and morphological operator demos in the rst docs files by @ramon-rd in https://github.com/kornia/kornia/pull/1884
* Image registration demo by @marianna13 in https://github.com/kornia/kornia/pull/1897
* [docs] Add total_variation_denoising gradio by @gagan3012 in https://github.com/kornia/kornia/pull/1880
* [Docs] Refactor the embedded Gradio demos by @NimaBoscarino in https://github.com/kornia/kornia/pull/1901

## RANSAC improvements

We have added homography-from-line-segments solver, as well as various speed-ups. We are not yet at OpenCV RANSAC quality level, more improvements to come :) But the line-solver is pretty unique! We also have example in our tutorials https://kornia-tutorials.readthedocs.io/en/latest/line_detection_and_matching_sold2.html

<img width="616" alt="image" src="https://user-images.githubusercontent.com/4803565/195350662-ca070a11-4c85-4082-b79d-c19eece5b328.png">

* Added homography from line segment correspondences by @ducha-aiki in https://github.com/kornia/kornia/pull/1851
* RANSAC improvements by @ducha-aiki in https://github.com/kornia/kornia/pull/1435
* Add get_perpendicular and get_closest_point_on_epipolar_line by @ducha-aiki in https://github.com/kornia/kornia/pull/1915
* Fix svdvals usage by @ducha-aiki in https://github.com/kornia/kornia/pull/1926

## Apple Silicon M1 support is closer, CI improvements

We are slowly working on being able to run kornia on M1. So far we have added possibility to test locally on M1 and mostly report Pytorch MPS backend crashes in various use-cases. Once this work is finished, we may provide some workarounds to have kornia-M1

* remove conv3d from spatial_gradient by @ducha-aiki in https://github.com/kornia/kornia/pull/1898
* Added possibility to run tests for mps locally by @ducha-aiki in https://github.com/kornia/kornia/pull/1716
* CI update pytorch-->-1.12.1 by @ducha-aiki in https://github.com/kornia/kornia/pull/1892
* lts is not supported on mac-os, separate it by @ducha-aiki in https://github.com/kornia/kornia/pull/1904
* Update setup to declarative metadata by @johnnv1 in https://github.com/kornia/kornia/pull/1885
* Bump pytest from 7.1.2 to 7.1.3 by @dependabot in https://github.com/kornia/kornia/pull/1860
* add concurrency cancel-in-progress in cpu workflow by @edgarriba in https://github.com/kornia/kornia/pull/1865
* Use --no-implicit-optional for type checking by @hauntsaninja in https://github.com/kornia/kornia/pull/1910

## Quaternion improvements

Implemented [Quaternion.slerp](https://github.com/kornia/kornia/blob/master/kornia/geometry/quaternion.py#L338) to interpolate between quaternions using quaternion arithmetic -- contributed by @cjpurackal 

```python
import torch
from kornia.geometry.quaternion import Quaternion

q0 = Quaternion.identity(batch_size=1)
q1 = Quaternion(torch.tensor([[1., .5, 0., 0.]]))
q2 = q0.slerp(q1, .3)

```

## More augmentations!

* [feat] Added Jigsaw Augmentation by @shijianjian in https://github.com/kornia/kornia/pull/1852
* [Feat] Added AugmentationDispatcher by @shijianjian https://github.com/kornia/kornia/pull/191

## What's Changed
* Added homography from line segment correspondences by @ducha-aiki in https://github.com/kornia/kornia/pull/1851
* Fixes typo bug that influences LoFTR training by @georg-bn in https://github.com/kornia/kornia/pull/1854
* Enlargen LoFTR positional encoding map if large images are input by @georg-bn in https://github.com/kornia/kornia/pull/1853
* docs: clarify the relation of color_jitter and color_jiggle by @kunaltyagi in https://github.com/kornia/kornia/pull/1858
* Bump pytest from 7.1.2 to 7.1.3 by @dependabot in https://github.com/kornia/kornia/pull/1860
* [pre-commit.ci] pre-commit suggestions by @pre-commit-ci in https://github.com/kornia/kornia/pull/1833
* Binary focal loss: Use pre-computed probs and increase readability by @klieret in https://github.com/kornia/kornia/pull/1848
* add concurrency cancel-in-progress in cpu workflow by @edgarriba in https://github.com/kornia/kornia/pull/1865
* modifying `add_weighted` to accept Tensors for `alpha`/`beta`/`gamma` by @nitaifingerhut in https://github.com/kornia/kornia/pull/1868
* Make AdaLAM output match confidence by @ducha-aiki in https://github.com/kornia/kornia/pull/1862
* fix bugs shift_rgb by @duc12111 in https://github.com/kornia/kornia/pull/1861
* Added gradio Image Stitching demo link by @kadirnar in https://github.com/kornia/kornia/pull/1871
* edge detection demo by @p-mishra1 in https://github.com/kornia/kornia/pull/1876
* Added Hugging Face edge detection demo link by @ramon-rd in https://github.com/kornia/kornia/pull/1874
* fix AdaLAM crash by @ducha-aiki in https://github.com/kornia/kornia/pull/1881
* [feat] Added Jigsaw Augmentation by @shijianjian in https://github.com/kornia/kornia/pull/1852
* [docs] add gradio app html and embeddings in filters by @lappemic in https://github.com/kornia/kornia/pull/1883
* created resize_antialias.html file by @gauthamk28 in https://github.com/kornia/kornia/pull/1877
* Adalam fix2 by @ducha-aiki in https://github.com/kornia/kornia/pull/1888
* Add edge detector and morphological operator demos in the rst docs files by @ramon-rd in https://github.com/kornia/kornia/pull/1884
* I added html file for module Line Fitting by @kadirnar in https://github.com/kornia/kornia/pull/1886
* No crash in local feature matching if empty tensor output by @ducha-aiki in https://github.com/kornia/kornia/pull/1890
* CI update pytorch-->-1.12.1 by @ducha-aiki in https://github.com/kornia/kornia/pull/1892
* Fix fail test by @ducha-aiki in https://github.com/kornia/kornia/pull/1896
* Added possibility to run tests for mps locally by @ducha-aiki in https://github.com/kornia/kornia/pull/1716
* remove conv3d from spatial_gradient by @ducha-aiki in https://github.com/kornia/kornia/pull/1898
* lts is not supported on mac-os, separate it by @ducha-aiki in https://github.com/kornia/kornia/pull/1904
* [pre-commit.ci] pre-commit suggestions by @pre-commit-ci in https://github.com/kornia/kornia/pull/1900
* quaternion index bug fix by @cjpurackal in https://github.com/kornia/kornia/pull/1903
* Update setup to declarative metadata by @johnnv1 in https://github.com/kornia/kornia/pull/1885
* [pre-commit.ci] pre-commit suggestions by @pre-commit-ci in https://github.com/kornia/kornia/pull/1912
* Image registration demo by @marianna13 in https://github.com/kornia/kornia/pull/1897
* [docs] Add total_variation_denoising gradio by @gagan3012 in https://github.com/kornia/kornia/pull/1880
* [Docs] Refactor the embedded Gradio demos by @NimaBoscarino in https://github.com/kornia/kornia/pull/1901
* Add get_perpendicular and get_closest_point_on_epipolar_line by @ducha-aiki in https://github.com/kornia/kornia/pull/1915
* Use --no-implicit-optional for type checking by @hauntsaninja in https://github.com/kornia/kornia/pull/1910
* Add Random Gamma and test by @duc12111 in https://github.com/kornia/kornia/pull/1837
* Add device query to Pinhole class by @YanivHollander in https://github.com/kornia/kornia/pull/1760
* Quaternion from axis angle representation  by @cjpurackal in https://github.com/kornia/kornia/pull/1917
* NeRF Implementation by @YanivHollander in https://github.com/kornia/kornia/pull/1911
* Bump pytest-mypy from 0.9.1 to 0.10.0 by @dependabot in https://github.com/kornia/kornia/pull/1919
* Geometry image transform demo by @dvando in https://github.com/kornia/kornia/pull/1922
* add spaces demo by @johko in https://github.com/kornia/kornia/pull/1905
* add space demo for homography warping by @johko in https://github.com/kornia/kornia/pull/1924
* RANSAC improvements by @ducha-aiki in https://github.com/kornia/kornia/pull/1435
* Fix warning in AdaLAM by @Skydes in https://github.com/kornia/kornia/pull/1925
* Fix svdvals usage by @ducha-aiki in https://github.com/kornia/kornia/pull/1926
* Bump pytest-cov from 3.0.0 to 4.0.0 by @dependabot in https://github.com/kornia/kornia/pull/1918
* fix shift_rgb stack dimension by @nmichlo in https://github.com/kornia/kornia/pull/1930
* Bump accelerate from 0.12.0 to 0.13.1 by @dependabot in https://github.com/kornia/kornia/pull/1937
* [pre-commit.ci] pre-commit suggestions by @pre-commit-ci in https://github.com/kornia/kornia/pull/1939
* [Fix] Fixed mypy warnings by @shijianjian in https://github.com/kornia/kornia/pull/1920
* Update kernels.py by @farhankhot in https://github.com/kornia/kornia/pull/1940
* Quaternion.norm bug fix by @cjpurackal in https://github.com/kornia/kornia/pull/1935
* [Feat] Added AugmentationDispatcher by @shijianjian in https://github.com/kornia/kornia/pull/1914
* Add `EdgeDetection` api by @edgarriba in https://github.com/kornia/kornia/pull/1483
* [feat] slerp implementation for Quaternion by @cjpurackal in https://github.com/kornia/kornia/pull/1931
* add laplacian pyramid by @lafith in https://github.com/kornia/kornia/pull/1816
* Fix quaternion doctests by @edgarriba in https://github.com/kornia/kornia/pull/1943
* Remove unnecessary CI jobs by @johnnv1 in https://github.com/kornia/kornia/pull/1933
* fix cuda tests failing by @ducha-aiki in https://github.com/kornia/kornia/pull/1941

## New Contributors
* @georg-bn made their first contribution in https://github.com/kornia/kornia/pull/1854
* @kunaltyagi made their first contribution in https://github.com/kornia/kornia/pull/1858
* @klieret made their first contribution in https://github.com/kornia/kornia/pull/1848
* @duc12111 made their first contribution in https://github.com/kornia/kornia/pull/1861
* @kadirnar made their first contribution in https://github.com/kornia/kornia/pull/1871
* @p-mishra1 made their first contribution in https://github.com/kornia/kornia/pull/1876
* @ramon-rd made their first contribution in https://github.com/kornia/kornia/pull/1874
* @lappemic made their first contribution in https://github.com/kornia/kornia/pull/1883
* @gauthamk28 made their first contribution in https://github.com/kornia/kornia/pull/1877
* @cjpurackal made their first contribution in https://github.com/kornia/kornia/pull/1903
* @johnnv1 made their first contribution in https://github.com/kornia/kornia/pull/1885
* @marianna13 made their first contribution in https://github.com/kornia/kornia/pull/1897
* @gagan3012 made their first contribution in https://github.com/kornia/kornia/pull/1880
* @hauntsaninja made their first contribution in https://github.com/kornia/kornia/pull/1910
* @dvando made their first contribution in https://github.com/kornia/kornia/pull/1922
* @johko made their first contribution in https://github.com/kornia/kornia/pull/1905
* @Skydes made their first contribution in https://github.com/kornia/kornia/pull/1925
* @nmichlo made their first contribution in https://github.com/kornia/kornia/pull/1930
* @farhankhot made their first contribution in https://github.com/kornia/kornia/pull/1940
* @lafith made their first contribution in https://github.com/kornia/kornia/pull/1816

**Full Changelog**: https://github.com/kornia/kornia/compare/v0.6.7...v0.6.8

<hr /><em>This discussion was created from the release <a href='https://github.com/kornia/kornia/releases/tag/v0.6.8'>v0.6.8 Experimental Nerf, Deep Edge Detection, community demos, Quaterion slerp, augmentations dispatcher, image matching and other bug fixes</a>.</em>
